src/Analyzers/CSharp/CodeFixes/ConvertNameOf/CSharpConvertNameOfCodeFixProvider.cs

```diff
@@ -0,0 +1,106 @@
+    {
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+        public CSharpConvertNameOfCodeFixProvider()
+        {
+        }
+
+        public override ImmutableArray<string> FixableDiagnosticIds
+           => ImmutableArray.Create(IDEDiagnosticIds.ConvertNameOfDiagnosticId);
+
+        internal sealed override CodeFixCategory CodeFixCategory => CodeFixCategory.CodeStyle;
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(new MyCodeAction(
+                c => FixAsync(context.Document, context.Diagnostics.First(), c)),
+               context.Diagnostics);
+            return Task.CompletedTask;
+        }
+
+        protected override async Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var isUsingSystem = root.
+                                DescendantNodes().
+                                OfType<UsingDirectiveSyntax>().
+                                Any(node => node.Name.ToString().Equals("System"));
+
+            foreach (var diagnostic in diagnostics)
+            {
+                var node = (MemberAccessExpressionSyntax)root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
+                ConvertNameOf(editor, node, semanticModel, isUsingSystem);
+            }
+        }
+
+        /**
+         * Method converts typeof(...).Name to nameof(...)
+         * The isUsingSystem parameter determines whether includes the System directive.
+         */
+        internal static void ConvertNameOf(
+            SyntaxEditor editor, MemberAccessExpressionSyntax node, SemanticModel semanticModel, bool isUsingSystem)
+        {
+            var exp = (TypeOfExpressionSyntax)node.Expression;
+            var idName = exp.Type.ToString();
+
+            //check if exp type is predefined type and convert
+            //example: int -> Int32, string -> String, etc
+            if (exp.Type.IsKind(SyntaxKind.PredefinedType))
+            {
+                idName = semanticModel.GetSymbolInfo(exp.Type).Symbol.
+                                       GetSymbolType().SpecialType.
+                                       ToPredefinedType().ToString();
+            }
+
+            //check if user is using System;
+            idName = isUsingSystem ? idName : "System." + idName;
+
+            var nameOfSyntax = InvocationExpression(IdentifierName("nameof")).
```

Comment:
instead of this, do `SyntaxGenerator.NameOfExpression()`  it will properly make the right sort of `nameof` token that the compielr understands.

Reply:
So when I use the SyntaxGenerator to create the nameof expression it simplifies the expression. For example, System.Int32 turns to int which won't compile. @CyrusNajmabadi